File: gateway/app/services/message.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, Optional


@dataclass
class MessageFlags:
    encrypted: bool = False
    qos: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        data: Dict[str, Any] = {}
        if self.encrypted:
            data["encrypted"] = True
        if self.qos:
            data["qos"] = self.qos
        return data


@dataclass
class InternalMessage:
    schema_version: int
    internal_id: str
    timestamp_ms: int
    user_id: str
    channel_id: str
    flags: MessageFlags
    payload: Any

    def to_client_payload(self) -> Dict[str, Any]:
        payload: Dict[str, Any] = {
            "id": self.internal_id,
            "ts": self.timestamp_ms,
            "channel_id": self.channel_id,
            "payload": self.payload,
        }
        if self.flags.encrypted:
            payload["encrypted"] = True
        if self.flags.qos:
            payload["qos"] = self.flags.qos
        return payload


def extract_channel_id(data: Dict[str, Any], fallback: str) -> str:
    for key in ("channel_id", "channel", "topic"):
        value = data.get(key)
        if isinstance(value, str) and value.strip():
            return value.strip()
    return fallback


def extract_payload(data: Dict[str, Any]) -> Any:
    if "payload" in data:
        return data.get("payload")
    return data


def extract_flags(data: Dict[str, Any]) -> MessageFlags:
    flags = MessageFlags()
    if isinstance(data.get("encrypted"), bool):
        flags.encrypted = bool(data.get("encrypted"))
    nested = data.get("flags")
    if isinstance(nested, dict) and isinstance(nested.get("encrypted"), bool):
        flags.encrypted = bool(nested.get("encrypted"))
    qos = data.get("qos")
    if isinstance(qos, str) and qos:
        flags.qos = qos
    if isinstance(nested, dict):
        qos = nested.get("qos")
        if isinstance(qos, str) and qos:
            flags.qos = qos
    return flags


def internal_from_dict(data: Dict[str, Any]) -> Optional[InternalMessage]:
    internal_id = data.get("internal_id")
    timestamp_ms = data.get("timestamp_ms")
    user_id = data.get("user_id")
    channel_id = data.get("channel_id")
    payload = data.get("payload")
    if not internal_id or not user_id or not channel_id:
        return None
    if not isinstance(timestamp_ms, int):
        return None
    flags_data = data.get("flags")
    flags = MessageFlags()
    if isinstance(flags_data, dict):
        if isinstance(flags_data.get("encrypted"), bool):
            flags.encrypted = bool(flags_data.get("encrypted"))
        qos = flags_data.get("qos")
        if isinstance(qos, str) and qos:
            flags.qos = qos
    return InternalMessage(
        schema_version=int(data.get("schema_version") or 1),
        internal_id=str(internal_id),
        timestamp_ms=timestamp_ms,
        user_id=str(user_id),
        channel_id=str(channel_id),
        flags=flags,
        payload=payload,
    )

File: rust_gateway/src/services/message.rs
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub(crate) struct MessageFlags {
    #[serde(default, skip_serializing_if = "is_false")]
    pub(crate) encrypted: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) qos: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub(crate) struct InternalMessage {
    #[serde(default = "schema_v1")]
    pub(crate) schema_version: u8,
    pub(crate) internal_id: String,
    pub(crate) timestamp_ms: i64,
    pub(crate) user_id: String,
    pub(crate) channel_id: String,
    #[serde(default)]
    pub(crate) flags: MessageFlags,
    pub(crate) payload: Value,
}

impl InternalMessage {
    pub(crate) fn to_client_payload(&self) -> Value {
        let mut payload = json!({
            "id": self.internal_id,
            "ts": self.timestamp_ms,
            "channel_id": self.channel_id,
            "payload": self.payload,
        });
        if let Some(map) = payload.as_object_mut() {
            if self.flags.encrypted {
                map.insert("encrypted".to_string(), Value::Bool(true));
            }
            if let Some(qos) = &self.flags.qos {
                map.insert("qos".to_string(), Value::String(qos.clone()));
            }
        }
        payload
    }
}

pub(crate) fn message_channel_id(data: &Value, fallback: &str) -> String {
    if let Some(value) = data.get("channel_id") {
        let id = value_to_string(value);
        if !id.is_empty() {
            return id;
        }
    }
    if let Some(value) = data.get("channel") {
        let id = value_to_string(value);
        if !id.is_empty() {
            return id;
        }
    }
    if let Some(value) = data.get("topic") {
        let id = value_to_string(value);
        if !id.is_empty() {
            return id;
        }
    }
    fallback.to_string()
}

pub(crate) fn message_payload(data: &Value) -> Value {
    data.get("payload").cloned().unwrap_or_else(|| data.clone())
}

pub(crate) fn message_flags(data: &Value) -> MessageFlags {
    let mut flags = MessageFlags::default();
    if let Some(value) = data.get("encrypted").and_then(|v| v.as_bool()) {
        flags.encrypted = value;
    }
    if let Some(value) = data
        .get("flags")
        .and_then(|v| v.get("encrypted"))
        .and_then(|v| v.as_bool())
    {
        flags.encrypted = value;
    }
    if let Some(value) = data.get("qos") {
        let qos = value_to_string(value);
        if !qos.is_empty() {
            flags.qos = Some(qos);
        }
    }
    if let Some(value) = data
        .get("flags")
        .and_then(|v| v.get("qos"))
    {
        let qos = value_to_string(value);
        if !qos.is_empty() {
            flags.qos = Some(qos);
        }
    }
    flags
}

fn is_false(value: &bool) -> bool {
    !*value
}

fn schema_v1() -> u8 {
    1
}

fn value_to_string(value: &Value) -> String {
    if let Some(s) = value.as_str() {
        return s.to_string();
    }
    if let Some(n) = value.as_i64() {
        return n.to_string();
    }
    if let Some(n) = value.as_u64() {
        return n.to_string();
    }
    if let Some(n) = value.as_f64() {
        let s = n.to_string();
        if !s.is_empty() {
            return s;
        }
    }
    String::new()
}
